### PR TITLE
AJ-1680 Use rawls models

### PR DESCRIPTION
**Pull request by calypsomatic**
https://github.com/DataBiosphere/terra-workspace-data-service/pull/631 failed to use the correct models for interaction with Rawls.  

In this PR, the `RawlsSnapshotSupport` now works with the models actually returned by Rawls, but translates them into the WSM models; this allows `RawlsSnapshotSupport` and `WsmSnapshotSupport` to continue to share code.

---

TODOs for future PRs:
- [ ] review use of RestTemplate vs. [RestClient](https://spring.io/blog/2023/07/13/new-in-spring-6-1-restclient). RestClient supports an ObservationRegistry.
- [ ] handle retries
- [ ] tests for retries
- [ ] maybe, combine the model classes needed by `RawlsClient` with the model classes needed by `RawlsRecordSink`

### CODE REVIEW

File: service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupport.java [hunk 1]
```diff
@@ -34,10 +47,90 @@ protected void linkSnapshot(UUID snapshotId) {
...
+  @Nullable
+  protected UUID safeGetSnapshotId(DataRepoSnapshotResource snapshotResource) {
+    DataRepoSnapshotAttributes dataRepoSnapshot = snapshotResource.getAttributes();
+    if (dataRepoSnapshot != null) {
+      String snapshotIdStr = dataRepoSnapshot.getSnapshot();
+      try {
+        return UUID.fromString(snapshotIdStr);
+      } catch (Exception e) {
+
+        String resourceId = "unknown";
```
Comment (calypsomatic) [inline service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupport.java:86]: This weird error logging has now been returned to the implementing `SnapshotSupport`s, so I could put the `workspaceId` back in.  It's also a weird bit of duplicated that could be pulled out to be reused, but not sure if it's worth it just to log a message.

File: service/src/test/java/org/databiosphere/workspacedataservice/dataimport/pfb/PfbQuartzJobControlPlaneE2ETest.java [hunk 2]
```diff
@@ -110,9 +112,13 @@ class PfbQuartzJobControlPlaneE2ETest {
   @BeforeEach
   void setup() {
     collectionId = UUID.randomUUID();
+
     // stub out rawls to report no snapshots already linked to this workspace
     when(rawlsClient.enumerateDataRepoSnapshotReferences(any(), anyInt(), anyInt()))
-        .thenReturn(new ResourceList());
+        .thenReturn(
+            new SnapshotListResponse(
+                CollectionConverters.asScala(new java.util.ArrayList<DataRepoSnapshotResource>())
```
Comment (calypsomatic) [inline service/src/test/java/org/databiosphere/workspacedataservice/dataimport/pfb/PfbQuartzJobControlPlaneE2ETest.java:120]: Really struggled with dealing with java-to-scala and back, am happy for any advice on how to just create a dummy response here.

File: service/src/test/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupportTest.java [hunk 3]
```diff
@@ -62,30 +63,89 @@ void paginateExistingSnapshots(int count) {
...
-            .toList();
+        actual.stream().map(res -> res.getAttributes().getInstanceName()).distinct().toList();
     assertEquals(count, snapshotInstanceNames.size());
     // assert the number of requests made to Rawls to generate the list
     double expectedInvocations = Math.floor((double) count / testPageSize) + 1;
     verify(rawlsClient, times((int) expectedInvocations))
         .enumerateDataRepoSnapshotReferences(any(), anyInt(), anyInt());
   }
+
+  @Test
```
Comment (calypsomatic) [inline service/src/test/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupportTest.java:85]: All these tests now mirror those in `WsmSnapshotSupport`, with minimal changes due to the different structure of `DataRepoSnapshotResource`.  Please check that the tests still make sense and that there isn't anything new I should be testing here.

File: service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupport.java [hunk 4]
```diff
@@ -25,6 +30,18 @@ public RawlsSnapshotSupport(
...
+  @Override
+  protected ResourceList enumerateDataRepoSnapshotReferences(int offset, int pageSize) {
+    RestClientRetry.RestCall<SnapshotListResponse> restCall =
+        (() -> rawlsClient.enumerateDataRepoSnapshotReferences(workspaceId.id(), offset, pageSize));
+    SnapshotListResponse snapshotListResponse =
+        restClientRetry.withRetryAndErrorHandling(
+            restCall, "Rawls.enumerateDataRepoSnapshotReferences");
+
+    return toResourceList(snapshotListResponse.gcpDataRepoSnapshots());
+  }
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupport.java:42]: apologies for the noise; I reorganized this class so methods are in the same order as WsmSnapshotSupport

File: service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/SnapshotSupport.java [hunk 5]
```diff
@@ -22,10 +22,11 @@
 
 public abstract class SnapshotSupport {
 
-  protected static final String DEFAULT_PRIMARY_KEY = "datarepo_row_id";
+  private static final String DEFAULT_PRIMARY_KEY = "datarepo_row_id";
```
Comment (davidangb) [inline service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/SnapshotSupport.java:25]: no real functional changes in this file, just a bunch of cleanup/documentation

File: service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupport.java [hunk 6]
```diff
@@ -34,10 +47,90 @@ protected void linkSnapshot(UUID snapshotId) {
...
+   */
+  protected List<DataRepoSnapshotResource> listAllSnapshots(int pageSize) {
+    int offset = 0;
+    final int hardLimit = 10000; // under no circumstances return more than this many snapshots
+
+    List<DataRepoSnapshotResource> finalList = new ArrayList<>(); // collect our results
+
+    while (offset < hardLimit) {
+      // get a page of results
+      SnapshotListResponse thisPage = enumerateDataRepoSnapshotReferences(offset, pageSize);
```
[the remark this answers is not in the document]
Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupport.java]: this comment is now obsolete

File: service/src/test/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupportTest.java (hunk 3 above)
Comment (davidangb) [reply service/src/test/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupportTest.java]: this comment is now obsolete

File: service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupport.java [hunk 7]
```diff
@@ -33,11 +50,36 @@ protected void linkSnapshot(UUID snapshotId) {
...
-        restCall, "Rawls.enumerateDataRepoSnapshotReferences");
+  /**
+   * Transforms a List<DataRepoSnapshotResource> to a ResourceList. RawlsSnapshotSupport and
+   * WsmSnapshotSupport return slightly different models from their API calls; performing this
+   * translation allows the two SnapshotSupport implementations to share code elsewhere.
+   *
+   * @param dataRepoSnapshotResourceList the model returned by Rawls APIs
+   * @return the model returned by WSM APIs
+   */
+  private ResourceList toResourceList(List<DataRepoSnapshotResource> dataRepoSnapshotResourceList) {
```
Comment (jladieu) [inline service/src/main/java/org/databiosphere/workspacedataservice/dataimport/snapshotsupport/RawlsSnapshotSupport.java:61]: Cool!

File: service/src/main/java/org/databiosphere/workspacedataservice/rawls/RawlsClient.java [hunk 8]
```diff
@@ -67,13 +67,19 @@ public void createSnapshotReference(UUID workspaceId, UUID snapshotId) {
...
           HttpMethod.POST,
-          new HttpEntity<>(new SnapshotModel().id(snapshotId), getAuthedHeaders()),
+          new HttpEntity<>(snapshotIdToNamedDataRepoSnapshot(snapshotId), getAuthedHeaders()),
           DataRepoSnapshotResource.class);
     } catch (RestClientResponseException e) {
       throw new RawlsException(e);
     }
   }
 
+  private NamedDataRepoSnapshot snapshotIdToNamedDataRepoSnapshot(UUID snapshotId) {
```
Comment (jladieu) [inline service/src/main/java/org/databiosphere/workspacedataservice/rawls/RawlsClient.java:77]: this is a totally optional code organization suggestion: would you consider putting this in a static method on `NamedDataRepoSnapshot` like:

`public static NamedDataRepoSnapshot fromSnapshotId(UUID snapshotId)`

?

The driving code organization principle I'm thinking here is trying to pull all creation patterns closer to the class definition, which makes it easier to scan all the various (and sometimes weird?) ways a class can get instantiated.

**Review by ashanhol — APPROVED** [review]

File: service/src/main/java/org/databiosphere/workspacedataservice/rawls/RawlsClient.java (hunk 8 above)
Comment (davidangb) [reply service/src/main/java/org/databiosphere/workspacedataservice/rawls/RawlsClient.java]: I agree, done

**Pull request merged**